File: src/main/java/com/assetmanagement/asset/entity/Asset.java
package com.assetmanagement.asset.entity;

import com.assetmanagement.asset.enums.AssetCategory;
import com.assetmanagement.asset.enums.AssetStatus;
import com.assetmanagement.common.entity.BaseEntity;
import com.assetmanagement.department.entity.Department;
import com.assetmanagement.employee.entity.Employee.Employee;
import com.assetmanagement.vendor.entity.Vendor;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;

@Entity
@Table(name = "assets")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Asset extends BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Version
    private Long version;

    @Column(nullable = false, unique = true)
    private String assetCode;

    @Column(nullable = false)
    private String assetName;

    @Column(nullable = false, unique = true)
    private String serialNumber;

    @Column(nullable = false)
    private String brand;

    @Column(nullable = false)
    private String model;

    @Column(length = 1000)
    private String description;

    @Column(nullable = false)
    private LocalDate purchaseDate;

    @Column(nullable = false, precision = 12, scale = 2)
    private BigDecimal purchaseCost;

    @Column(nullable = false)
    private LocalDate warrantyExpiry;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private AssetStatus status;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private AssetCategory category;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "department_id", nullable = false)
    private Department department;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_employee_id")
    private Employee assignedEmployee;

    // nullable = false not set here so that existing assets without a vendor don't fail.
    // New asset creation enforces vendorId at the service layer.
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "vendor_id")
    private Vendor vendor;
}

File: frontend/src/types/asset.ts
export type AssetStatus = "AVAILABLE" | "ASSIGNED" | "UNDER_MAINTENANCE" | "RETIRED";

export type AssetCategory = "LAPTOP" | "DESKTOP" | "MONITOR" | "PRINTER" | "MOBILE" | "TABLET" | "NETWORK_DEVICE" | "OTHER";

export interface AssetResponse {
  id: number;
  assetCode: string;
  assetName: string;
  serialNumber: string;
  brand: string;
  model: string;
  description: string;
  purchaseDate: string; // ISO date string
  purchaseCost: number;
  warrantyExpiry: string; // ISO date string
  status: AssetStatus;
  category: AssetCategory;
  departmentId: number;
  departmentName: string;
  assignedEmployeeId: number | null;
  assignedEmployeeName: string | null;
  vendorId: number;
  vendorName: string;
}

export interface AssetRequest {
  assetName: string;
  serialNumber: string;
  brand: string;
  model: string;
  description?: string;
  purchaseDate: string;
  purchaseCost: number;
  warrantyExpiry: string;
  category: AssetCategory;
  departmentId: number;
  assignedEmployeeId?: number | null;
  vendorId?: number; // Required for create, excluded for update
}
